File: Example/VRMaskColoredButton_ExampleObjC/ViewController.m
//
//  ViewController.m
//  VRMaskColoredButton_ExampleObjC
//
//  Created by Ivan Rublev on 2/1/16.
//  Copyright © 2016 Ivan Rublev http://ivanrublev.me. All rights reserved.
//

#import "ColorOptions.h"
#import "Jukebox.h"
#import "SegmentedView.h"
#import "ViewController.h"
@import VRMaskColoredButton;

@interface ViewController () <JukeboxDelegate>
@property (nonatomic) Jukebox *jukebox;

@property (strong, nonatomic) IBOutlet VRMaskColoredButton *muteButton;
@property (strong, nonatomic) IBOutlet UILabel *songTitle;

@property (strong, nonatomic) IBOutlet VRMaskColoredButton *backwardButton;
@property (strong, nonatomic) IBOutlet VRMaskColoredButton *playButton;
@property (strong, nonatomic) IBOutlet VRMaskColoredButton *forwardButton;
@property (strong, nonatomic) IBOutletCollection(VRMaskColoredButton) NSArray *playbackControls;
@property (strong, nonatomic) IBOutlet SegmentedView *segmentedView;

@property (nonatomic) ColorOption selectedColorOption;
@end

@implementation ViewController

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if ((self = [super initWithCoder:aDecoder])) {
        self.jukebox = [Jukebox new];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [[VRMaskColoredButton appearance] setNormalColor:buttonIconColor]; // Configure every button's image color via appearance proxy.
    
    [self setButtonTitleAndTag:self.segmentedView.segLeftButton withColorOption:ColorOptionPink];
    [self setButtonTitleAndTag:self.segmentedView.segMidButton withColorOption:ColorOptionGreen];
    [self setButtonTitleAndTag:self.segmentedView.segRightButton withColorOption:ColorOptionYellow];
    
    self.selectedColorOption = ColorOptionPink;
    self.jukebox.delegate = self;
}

- (void)setButtonTitleAndTag:(UIButton *)button withColorOption:(ColorOption)option {
    button.tag = option;
    [button setTitle:titleForColorOption[@(option)] forState:UIControlStateNormal];
}


#pragma mark -
#pragma mark Changing the color of playback controls
- (void)setSelectedColorOption:(ColorOption)selectedColorOption {
    _selectedColorOption = selectedColorOption;
    
    // Update the selected states of segmented buttons
    for (VRMaskColoredButton *segButton in self.segmentedView.segments) {
        segButton.selected = (segButton.tag == selectedColorOption);
    }
    // Change the color of playback buttons background image according to selected color option.
    UIColor *buttonBackgroundColor = colorForColorOption[@(selectedColorOption)];
    for (VRMaskColoredButton *playbackButton in self.playbackControls) {
        [playbackButton setBackgroundImageColor:buttonBackgroundColor forState:UIControlStateNormal];
        // Make the button's image and backround image colors to exchange in highlighted state.
        [playbackButton setBackgroundImageColor:buttonIconColor forState: UIControlStateHighlighted];
        [playbackButton setImageColor:buttonBackgroundColor forState: UIControlStateHighlighted];
        [playbackButton setBackgroundImageColor:buttonIconColor forState: UIControlStateSelected | UIControlStateHighlighted];
        [playbackButton setImageColor:buttonBackgroundColor forState: UIControlStateSelected | UIControlStateHighlighted];
    }
}

- (IBAction)segButtonPressed:(UIButton *)sender {
    self.selectedColorOption = (ColorOption)sender.tag;
}


#pragma mark -
#pragma mark Control jukebox
- (IBAction)controlButtonPressed:(UIButton *)sender {
    if (sender == self.playButton) {
        self.jukebox.play = !self.jukebox.play;
        
    } else if (sender == self.backwardButton) {
        [self.jukebox prevSong];
        
    } else if (sender == self.forwardButton) {
        [self.jukebox nextSong];
        
    } else if (sender == self.muteButton) {
        self.jukebox.muted = !self.jukebox.muted;
        
    } else {
        NSAssert(YES, @"Unhandled control");
    }
}

- (void)jukeboxPlayWasStarted {
    self.playButton.selected = YES;
    [self updateSongTitle];
}

- (void)jukeboxPlayWasPaused {
    self.playButton.selected = NO;
    [self updateSongTitle];
}

- (void)jukeboxSongWasChanged:(NSUInteger)currentSong {
    self.backwardButton.enabled = currentSong > 0;
    self.forwardButton.enabled = currentSong < self.jukebox.songTitles.count-1;
    [self updateSongTitle];
}

- (void)jukeboxVolume:(BOOL)muted {
    self.muteButton.selected = muted;
}

- (void)updateSongTitle {
    NSString *text = [NSString stringWithFormat:@"%lu. %@", self.jukebox.currentSong+1, self.jukebox.songTitles[self.jukebox.currentSong]];
    if (self.jukebox.play) {
        text = [text stringByAppendingString:@" *"];
    }
    self.songTitle.text = text;
}

@end

File: Example/VRMaskColoredButton_ExampleSwift/ViewController.swift
//
//  ViewController.swift
//  VRMaskColoredButton
//
//  Created by Ivan Rublev on 1/21/16.
//  Copyright (c) 2016 Ivan Rublev. All rights reserved.
//

import UIKit
import VRMaskColoredButton

class ViewController: UIViewController, JukeboxDelegate {

    private let jukebox = Jukebox()
    
    @IBOutlet var muteButton: VRMaskColoredButton!
    @IBOutlet var songTitle: UILabel!

    // Playback controls
    @IBOutlet var backwardButton: VRMaskColoredButton!
    @IBOutlet var playButton: VRMaskColoredButton!
    @IBOutlet var forwardButton: VRMaskColoredButton!
    @IBOutlet var playbackControls: [VRMaskColoredButton]!

    @IBOutlet var segmentedView: SegmentedView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        VRMaskColoredButton.appearance().normalColor = buttonIconColor // Configure every button's image color via appearance proxy.
        
        setButtonTitleAndTag(segmentedView.segLeftButton, withColorOption: .Pink)
        setButtonTitleAndTag(segmentedView.segMidButton, withColorOption: .Green)
        setButtonTitleAndTag(segmentedView.segRightButton, withColorOption: .Yellow)
        selectedColorOption = .Pink
        jukebox.delegate = self
    }
    
    private func setButtonTitleAndTag(button: UIButton, withColorOption option: ColorOption) {
        button.tag = option.rawValue
        button.setTitle(titleForColorOption[option]!, forState: .Normal)
    }

    // MARK: Changing the color of playback controls
    var selectedColorOption: ColorOption! {
        didSet {
            // Update the selected states of segmented buttons
            for segButton in segmentedView.segments {
                segButton.selected = (segButton.tag == selectedColorOption.rawValue)
            }
            // Change the color of playback buttons background image according to selected color option.
            let buttonBackgroundColor = colorForColorOption[selectedColorOption]
            for playbackButton in playbackControls {
                playbackButton.setBackgroundImageColor(buttonBackgroundColor, forState: .Normal)
                // Make the button's image and backround image colors to exchange in highlighted state.
                playbackButton.setBackgroundImageColor(buttonIconColor, forState: .Highlighted)
                playbackButton.setImageColor(buttonBackgroundColor, forState: .Highlighted)
                playbackButton.setBackgroundImageColor(buttonIconColor, forState: UIControlState.Selected.union(.Highlighted))
                playbackButton.setImageColor(buttonBackgroundColor, forState: UIControlState.Selected.union(.Highlighted))
            }
        }
    }

    @IBAction func segButtonPressed(sender: UIButton) {
        selectedColorOption = ColorOption(rawValue: sender.tag)
    }
    
    // MARK: Control jukebox
    @IBAction func controlButtonPressed(sender: UIButton) {
        switch sender {
        case playButton:
            jukebox.play = !jukebox.play
            
        case backwardButton:
            jukebox.prevSong()
            
        case forwardButton:
            jukebox.nextSong()
        
        case muteButton:
            jukebox.muted = !jukebox.muted
            
        default:
            assertionFailure("Unhandled control")
        }
    }
    
    func jukeboxPlayWasStarted() {
        playButton.selected = true
        updateSongTitle()
    }
    
    func jukeboxPlayWasPaused() {
        playButton.selected = false
        updateSongTitle()
    }
    
    func jukeboxSongWasChanged(currentSong: UInt) {
        backwardButton.enabled = currentSong > 0
        forwardButton.enabled = currentSong < UInt(jukebox.songTitles.count-1)
        updateSongTitle()
    }
    
    func jukeboxVolume(muted: Bool) {
        muteButton.selected = muted
    }

    func updateSongTitle() {
        var text = "\(jukebox.currentSong+1). \(jukebox.songTitles[Int(jukebox.currentSong)])"
        if jukebox.play {
            text += " *"
        }
        songTitle.text = text
    }
}
